### PR TITLE
Domains import pending transfers

### PR DESCRIPTION
**Pull request by wolfgangmeyers**
Depends on https://github.com/google/nomulus/pull/56 and https://github.com/google/nomulus/pull/52

### CODE REVIEW

**Comment by wolfgangmeyers** [comment]
Once domains import mapper is merged I will rebase on the latest master.

**Comment by CydeWeys** [comment]
Can you rebase this against HEAD and squash only the changes relative to HEAD into a single commit?  Thanks.

**Comment by wolfgangmeyers** [comment]
@CydeWeys updated.

**Comment by wolfgangmeyers** [comment]
I'm seeing a lot of weird changes in imports. Going to go through and clean up...

**Comment by CydeWeys** [comment]
The import order rules are very simple: sorted static imports all in one group, then a newline, then sorted normal imports all in one group.

**Comment by CydeWeys** [comment]
It looks like you've combined new functionality along with refactoring existing helper methods into their own classes.  That makes this harder to review, since I'm not seeing good diffs on the functions that were simply refactored, but then maybe also changed.  Can you split this into two PRs, one for the refactoring (that will be a rubber-stamp approval), and then a second one that handles import of pending transfers?  Thanks.

**Comment by wolfgangmeyers** [comment]
@CydeWeys see https://github.com/google/nomulus/pull/61

**Comment by CydeWeys** [comment]
The refactor is merged.   Please rebase to HEAD.

**Comment by wolfgangmeyers** [comment]
@CydeWeys rebase complete

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 1]
```diff
@@ -146,10 +256,13 @@ private void checkHistoryEntry(HistoryEntry entry, DomainResource parent) {
     assertThat(entry.getParent()).isEqualTo(Key.create(parent));
   }
 
-  /** Verifies that domain id and ROID match expected values */
+  /** Verifies that domain fields match expected values */
   private void checkDomain(DomainResource domain) {
     assertThat(domain.getFullyQualifiedDomainName()).isEqualTo("example1.test");
     assertThat(domain.getRepoId()).isEqualTo("Dexample1-TEST");
+    // TODO: really test these values
```
Comment (CydeWeys) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Did you mean to tackle this TODO now, or is there a follow-up PR?

Comment (wolfgangmeyers) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Good catch. I think I left this out, let me get back to you :)

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 2]
```diff
@@ -66,10 +83,10 @@ public RdeImportUtils(
     this.escrowBucketName = escrowBucketName;
   }
 
-  private <T extends EppResource> void importEppResource(final T resource, final String type) {
+  public <T extends EppResource> ImmutableSet<Object> getIndexesForEppResource(T resource, String type) {
```
Comment (mindhog) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Line too long.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java [hunk 3]
```diff
@@ -293,16 +245,33 @@ private static HistoryEntry createHistoryEntry(XjcRdeDomain domain) {
   }
 
   /** Converts {@link XjcRdeDomainTransferDataType} to {@link TransferData}. */
-  private static TransferData convertDomainTransferData(XjcRdeDomainTransferDataType data) {
+  private static TransferData convertDomainTransferData(XjcRdeDomainTransferDataType data, DateTime domainExpiration) {
```
Comment (mindhog) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: Line too long.

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 4]
```diff
@@ -135,6 +151,100 @@ public void test_mapreducePublishesToDns() throws Exception {
...
+          && clientId.equals(message.getClientId())
+          && expectedAt.equals(message.getEventTime())) {
+        return;
+      }
+    }
+    fail("Expected transfer request poll message");
+  }
+
+  /** Verifies the existence of a transfer request poll message */
+  private void checkTransferServerApprovalPollMessage(DomainResource domain, String clientId, DateTime expectedAt) {
```
Comment (mindhog) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Line too long.

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 5]
```diff
@@ -174,6 +287,14 @@ private void pushToGcs(ByteSource source) throws IOException {
     return result;
   }
 
+  private static BillingEvent.Recurring loadAutorenewBillingEventForDomain(DomainResource domain) {
```
Comment (mindhog) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Line too long.

Comment (mindhog) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Actually, nevermind.  It's not.

File: javatests/google/registry/rde/imports/XjcToDomainResourceConverterTest.java [hunk 6]
```diff
@@ -44,26 +48,29 @@
...
 import google.registry.testing.InjectRule;
 import google.registry.util.StringGenerator;
 import google.registry.xjc.rdedomain.XjcRdeDomain;
 import google.registry.xjc.rdedomain.XjcRdeDomainElement;
+import org.joda.time.DateTime;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
```
Comment (mindhog) [inline javatests/google/registry/rde/imports/XjcToDomainResourceConverterTest.java]: Why move these?  We typically order imports by sort order.

File: javatests/google/registry/rde/imports/XjcToDomainResourceConverterTest.java [hunk 7]
```diff
@@ -358,11 +365,62 @@ public void testConvertDomainResourceAutoRenewPollMessage() throws Exception {
...
+    final HistoryEntry historyEntry = createHistoryEntryForDomainImport(xjcDomain);
+    final BillingEvent.Recurring autorenewBillingEvent = createAutoRenewBillingEventForDomainImport(xjcDomain, historyEntry);
+    final PollMessage.Autorenew autorenewPollMessage = createAutoRenewPollMessageForDomainImport(xjcDomain, historyEntry);
     return ofy().transact(new Work<DomainResource>() {
+      @SuppressWarnings("unchecked")
       @Override
       public DomainResource run() {
-        return XjcToDomainResourceConverter.convertDomain(xjcDomain);
+        ofy().save().entities(historyEntry, autorenewBillingEvent, autorenewPollMessage);
+        return XjcToDomainResourceConverter.convertDomain(xjcDomain, autorenewBillingEvent, autorenewPollMessage);
```
Comment (mindhog) [inline javatests/google/registry/rde/imports/XjcToDomainResourceConverterTest.java]: Line too long.

**Comment by wolfgangmeyers** [comment]
@CydeWeys @mindhog all comments addressed

**Comment by mindhog** [comment]
LGTM

File: java/google/registry/rde/imports/RdeDomainImportAction.java [hunk 8]
```diff
@@ -170,6 +239,13 @@ public void vrun() {
         throw new DomainImportException(xjcDomain.getName(), xjcDomain.toString(), e);
       }
     }
+
+    private void setupPendingTransfer(final HistoryEntry historyEntry,
+        DomainResource domain,
+        BillingEvent.Recurring autorenewBillingEvent,
+        PollMessage.Autorenew autorenewPollMessage) {
+
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeDomainImportAction.java]: What's going on here?

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 9]
```diff
@@ -66,10 +83,11 @@ public RdeImportUtils(
     this.escrowBucketName = escrowBucketName;
   }
 
-  private <T extends EppResource> void importEppResource(final T resource, final String type) {
+  public <T extends EppResource> ImmutableSet<Object> getIndexesForEppResource(T resource,
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Use ImmutableObject, not Object, as the return type.

This needs a Javadoc.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 10]
```diff
@@ -83,9 +101,15 @@ public RdeImportUtils(
         type,
         resource.getForeignKey(),
         resource.getRepoId());
-    ofy.save().entity(resource);
-    ofy.save().entity(ForeignKeyIndex.create(resource, resource.getDeletionTime()));
-    ofy.save().entity(EppResourceIndex.create(Key.create(resource)));
+    return ImmutableSet.<Object>of(ForeignKeyIndex.create(resource, resource.getDeletionTime()),
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: This method is named "get" but it's really creating the two types of indexes.  Should the method instead be named "create"?

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 9 above)
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: This is typed with T extends EppResource, but it will do the wrong thing for DomainApplications, which do not have ForeignKeyIndexes.  Either check the type of T and throw an IllegalArgumentException if a domain application is passed, or use the more restrictive type of <T extends EppResource & ForeignKeyedEppResource> or similar.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 11]
```diff
@@ -83,9 +101,15 @@ public RdeImportUtils(
...
+        EppResourceIndex.create(Key.create(resource)));
+  }
+
+  private <T extends EppResource> void importEppResource(final T resource, final String type) {
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Same comments regarding generic type parameter and also the "type" method parameter.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 12]
```diff
@@ -66,10 +83,11 @@ public RdeImportUtils(
     this.escrowBucketName = escrowBucketName;
   }
 
-  private <T extends EppResource> void importEppResource(final T resource, final String type) {
+  public <T extends EppResource> ImmutableSet<Object> getIndexesForEppResource(T resource,
+      String type) {
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Why is type being passed in as a String?  What is it exactly, a class name?  It seems to me like you don't even need to pass it, since you can get what you need based on the class that resource is an instance of.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 13]
```diff
@@ -66,10 +83,11 @@ public RdeImportUtils(
     this.escrowBucketName = escrowBucketName;
   }
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: A lot of these methods could be made static, no?

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 14]
```diff
@@ -174,6 +331,14 @@ private void pushToGcs(ByteSource source) throws IOException {
...
```
Comment (CydeWeys) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: I'm wondering if these should be annotated with @Nullable, because ofy().load() will return null if the entity in question doesn't exist.  Now granted, it should always exist ... but it's more likely not to if someone failed in an import mapreduce, versus a standard transactional EPP flow, for which it's really hard to imagine such an occurrence ever happening.

**Comment by CydeWeys** [comment]
Ping?

**Comment by wolfgangmeyers** [comment]
Sorry, was on vacation for a few days. I'm going to jump on this now.

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java (hunk 14 above)
Comment (wolfgangmeyers) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Do we care about `@Nullable` in unit tests?

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 13 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: true. That would introduce more changes that are unrelated to the scope of this PR, let me know if you feel strongly about it :)

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 9 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: If I do that, I just need to convert the set later on to use Object. Instances of TransferServerApproveEntity are added to this list later on if a pending transfer is in progress. Changing this from Object to ImmutableObject will complicate code elsewhere and make it less readable.

Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Note https://github.com/google/nomulus/pull/57/files#diff-3385ce686547e7ce6c33380a1672d441R181 and https://github.com/google/nomulus/pull/57/files#diff-3385ce686547e7ce6c33380a1672d441R217

**Comment by wolfgangmeyers** [comment]
@CydeWeys all feedback responded to or implemented.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 15]
```diff
@@ -100,7 +123,7 @@ public RdeImportUtils(
    * created.
    */
   public void importHost(final HostResource resource) {
-    importEppResource(resource, "host");
+    importEppResource(resource);
```
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: These three methods seem unnecessary if all they're doing is calling the same generic-typed method.  Just call importEppResource() everywhere?

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java (hunk 14 above)
Comment (CydeWeys) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Why not?  It's a helpful annotation in all circumstances.  Here, it'll make it easier to write more tests, since the behavior of these helper methods is better documented.

**Comment by wolfgangmeyers** [comment]
@CydeWeys both fixed.

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 13 above)
Comment (CydeWeys) [inline java/google/registry/rde/imports/RdeImportUtils.java]: OK, we'll see what happens when we do the import.  They might end up becoming static anyway if our static code analysis tool ends up complaining.

**Comment by CydeWeys** [comment]
This is going through internal review now.

**Comment by CydeWeys** [comment]
This has been merged internally and will be included in our next sync.

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 16]
```diff
@@ -184,4 +179,59 @@ public static Trid generateTridForImport() {
...
+  }
+
+  public static HistoryEntry createHistoryEntryForDomainImport(XjcRdeDomain domain) {
+    XjcRdeDomainElement element = new XjcRdeDomainElement(domain);
+    final HistoryEntry historyEntry =
+        new HistoryEntry.Builder()
+            .setType(HistoryEntry.Type.RDE_IMPORT)
+            .setClientId(domain.getClID())
+            .setTrid(generateTridForImport())
+            .setModificationTime(DateTime.now())
```
Comment (nfelt) [inline java/google/registry/rde/imports/RdeImportUtils.java]: HistoryEntry modification times should always match the transaction time of the save - it makes it easier if need be to match the HistoryEntry log with other stuff that reflects the transaction time, including CreateAutoTimestamp/UpdateAutoTimestamp and the commit log system.  (In fact, HistoryEntry's modificationTime should arguably be a CreateAutoTimestamp; it predates the latter's existence.)

So here we should really pass in DateTime now as a parameter and provide ofy().getTransactionTime() from the callsite.

FWIW, also, in general our approach to time means it's typically a red flag to be using DateTime.now() except if you're explicitly not in a transaction context at all (like DomainInfoFlow).

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java [hunk 17]
```diff
@@ -181,14 +175,10 @@ public GracePeriod apply(XjcRgpStatusType gracePeriodStatus) {
...
-    BillingEvent.Recurring autoRenewBillingEvent =
-        createAutoRenewBillingEvent(domain, historyEntry);
-    PollMessage.Autorenew pollMessage = createAutoRenewPollMessage(domain, historyEntry);
-    ofy().save().<ImmutableObject>entities(historyEntry, autoRenewBillingEvent, pollMessage);
+  static DomainResource convertDomain(
+      XjcRdeDomain domain,
+      BillingEvent.Recurring autoRenewBillingEvent,
+      PollMessage.Autorenew autoRenewPollMessage) {
     GracePeriodConverter gracePeriodConverter =
         new GracePeriodConverter(domain, Key.create(autoRenewBillingEvent));
```
Comment (nfelt) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: This comment really applies more specifically to above where GracePeriodConverter.apply() calls GracePeriod.createForRecurring(), but GitHub apparently doesn't let you comment outside diff ranges.

I think using createForRecurring() in this context is incorrect.  It's being called with the Recurring billing event that we're passing in to this function, which we just newly created via createAutoRenewBillingEventForDomainImport().  The problem is, that Recurring event has an eventTime set to the domain's expiration time, which must be in the future relative to the import time.  Any given Recurring event is only expanded (e.g. by the ExpandRecurringBillingEventsAction) within the time range defined by [eventTime, recurrenceEndTime).  So the effect is that the grace period, if invoked, will trigger a cancellation that doesn't correspond to any actual billing event -  the cancellation will be for an autorenew that happened in the past, but the recurring event we're using doesn't start recurring until the future.

I'm curious more broadly about what you're expecting to do about billing for grace periods that span the migration from one TLD to another.  None of the other grace periods in GracePeriodConverter are being given billing events, so they won't generate any cancellations if the domain is deleted while still in the grace period.  Are you thinking that you'd just do some kind of one-time manual process to inspect for any domains deleted shortly after the migration, determine what grace periods they would have triggered, and issue refunds appropriately?

At any rate, while it's possible to hack around the autorenew issue (e.g. by setting the eventTime of the autorenew to the creationTime of the domain plus 1 year, which still may not be historically accurate but at least fixes the dangling cancellation problem), it seems like any reconciliation process you have would be easier to manage if it applied to all types of grace periods, rather than trying to special-case autorenews to do something cleverer.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java [hunk 18]
```diff
@@ -293,16 +238,34 @@ private static HistoryEntry createHistoryEntry(XjcRdeDomain domain) {
   }
 
   /** Converts {@link XjcRdeDomainTransferDataType} to {@link TransferData}. */
-  private static TransferData convertDomainTransferData(XjcRdeDomainTransferDataType data) {
+  private static TransferData convertDomainTransferData(XjcRdeDomainTransferDataType data,
+      DateTime domainExpiration) {
     if (data == null) {
       return TransferData.EMPTY;
     }
+    // If the transfer is pending, calculate the number of years to add to the domain expiration
```
Comment (nfelt) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: This calculation will need to subsume an autorenew's extra year if the domain will still be within the autorenew grace period at the automatic transfer time.  However, there are existing issues with how our transfer flows implement this, so I will take on figuring out a consistent fix for both the EPP flows and here - this is just FYI.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java [hunk 19]
```diff
@@ -293,16 +238,34 @@ private static HistoryEntry createHistoryEntry(XjcRdeDomain domain) {
...
+    // on approval of the transfer.
+    TransferStatus transferStatus = TRANSFER_STATUS_MAPPER.xmlToEnum(data.getTrStatus().value());
+    // Get new expiration date
+    DateTime newExpirationTime = domainExpiration;
+    if (transferStatus == TransferStatus.PENDING) {
+      // Default to domain expiration time plus one year if no expiration is specified
```
Comment (nfelt) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: Do you anticipate getting deposits where there is a pending transfer yet no exDate specified?  I'll admit the specs are fairly vague about exDate and do indicate it's optional, but I'm curious to know whether this special-casing is in response to deposits you've seen "in the wild" or just a contingency.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java [hunk 20]
```diff
@@ -293,16 +238,34 @@ private static HistoryEntry createHistoryEntry(XjcRdeDomain domain) {
...
+      }
+    }
     return new TransferData.Builder()
-        .setTransferStatus(TRANSFER_STATUS_MAPPER.xmlToEnum(data.getTrStatus().value()))
+        .setTransferStatus(transferStatus)
         .setGainingClientId(data.getReRr().getValue())
         .setLosingClientId(data.getAcRr().getValue())
         .setTransferRequestTime(data.getReDate())
         .setPendingTransferExpirationTime(data.getAcDate())
+        // This will be wrong for domains that are not in pending transfer,
```
Comment (nfelt) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: Our extendedRegistrationYears field has a number of problems overall, not the least of which is that as far as I can tell, ICANN policy stipulates that transfers must include a fixed 1-year registration extension, not an arbitrary number of years.  See section A.8 of ICANN's ["Policy on Transfer of Registrations between Registrars"](https://www.icann.org/resources/pages/policy-2012-03-07-en) (note that regular transfers are covered under section A, "Holder-authorized transfers"; section B handles "ICANN-Approved transfers" which are bulk transfers between one registrar to another, e.g. upon going out of business or being acquired).

> 8. Effect on Term of Registration
> The completion by Registry Operator of a holder-authorized transfer under this Part A shall result in a one-year extension of the existing registration, provided that in no event shall the total unexpired term of a registration exceed ten (10) years.

At least, that's the interpretation that seems most straightforward.  While EPP in general and our registry implementation to date allow specifying a number >1 of years to renew in a transfer (see [the tests](https://github.com/google/nomulus/blob/master/javatests/google/registry/flows/domain/testdata/domain_transfer_request_response_10_years.xml)), evidently that's not the same as the policy.  And in practice it appears that registrars only do 1-year transfers.

So I'm hoping to fix that and in the process eliminate this field, which will eliminate this issue.

However, all that said, I'm pointing out just FYI that this calculation isn't even always correct for domains that are in pending transfer, due to 10-year capping.  As mentioned in my other comment, if the transfer would hit 10-year capping (e.g. the domain was created w/ a 10 year registration, next day a transfer is requested w/ bundled 1-year registration extension), then exDate won't necessarily be a full year from the current expiration time, and the Years.yearsBetween() call will truncate and return 0, which is never a valid value of extendedRegistrationYears.

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 21]
```diff
@@ -135,6 +148,151 @@ public void test_mapreducePublishesToDns() throws Exception {
...
+        DateTime.parse("2015-04-03T22:00:00.0Z"),
+        DateTime.parse("2015-01-08T22:00:00.0Z"));
+    checkAutorenewPollMessage(beforeApproval,
+        "RegistrarX",
+        DateTime.parse("2015-04-03T22:00:00.0Z"),
+        DateTime.parse("2015-01-08T22:00:00.0Z"));
+
+    // Domain should be assigned to RegistrarY after server approval
+    DomainResource afterApproval =
+        domains.get(0).cloneProjectedAtTime(serverApprovalTime.plus(Seconds.ONE));
```
Comment (nfelt) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: No need to add the extra second; the transfer is effective precisely at the server approval time, inclusive.  See e.g. the test here:
https://github.com/google/nomulus/blob/ec55aa5361f71d86682817252054cbaf2caf9c7c/javatests/google/registry/flows/domain/DomainTransferRequestFlowTest.java#L242

Better to be precise to avoid giving the impression that the boundary is fuzzy.

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java [hunk 22]
```diff
@@ -135,6 +148,151 @@ public void test_mapreducePublishesToDns() throws Exception {
     assertDnsTasksEnqueued("example1.test");
   }
 
+  /**
+   * Verifies the state of an imported pending transfer before and after implicit server approval
+   */
+  @Test
+  public void testMapreducePendingTransferServerApproval() throws Exception {
```
Comment (nfelt) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: FWIW, there are additional tests that would be good to have either here or in XjcToDomainResourceConverterTest, specifically:

- autorenew grace period subsuming behavior (I can take care of this one)
- 10-year capping case, where transfer exDate isn't a year multiple after current exDate
- domain with previous (non-pending) transfer data - domain_fragment_transferPeriod.xml actually includes this, but the TransferData fields themselves are not checked in the test, only the grace period

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 23]
```diff
@@ -66,10 +81,11 @@ public RdeImportUtils(
     this.escrowBucketName = escrowBucketName;
   }
 
-  private <T extends EppResource> void importEppResource(final T resource, final String type) {
+  public <T extends EppResource & ForeignKeyedEppResource> ImmutableSet<Object>
```
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: I deliberately avoided this, because I have to store them in another collection that has objects that are represented as interfaces only. If I change it here I would need to do typecasting elsewhere.

**Comment by wolfgangmeyers** [comment]
@nfelt I'm currently busy with something else but I will come back to this when I have time.

File: java/google/registry/rde/imports/RdeDomainImportAction.java [hunk 24]
```diff
@@ -145,17 +161,70 @@ private RdeImportUtils createRdeImportUtils() {
...
+              Money transferCost = getDomainRenewCost(
+                  domain.getFullyQualifiedDomainName(),
+                  transferData.getPendingTransferExpirationTime(),
+                  transferData.getExtendedRegistrationYears());
+              // Create speculative entities in anticipation of an automatic server approval.
+              ImmutableSet<TransferServerApproveEntity> serverApproveEntities =
+                  createTransferServerApproveEntities(
+                      transferData.getPendingTransferExpirationTime(),
+                      domain.getRegistrationExpirationTime()
+                          .plusYears(transferData.getExtendedRegistrationYears()),
```
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeDomainImportAction.java]: Looks like you found a bug. `transferData.getPendingTransferExpirationTime()` should not be used below, I need to calculate the new expiration using `extendRegistrationWithCap(..)`

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 25]
```diff
@@ -66,10 +81,11 @@ public RdeImportUtils(
...
+      createIndexesForEppResource(T resource) {
     @SuppressWarnings("unchecked")
     Class<T> resourceClass = (Class<T>) resource.getClass();
-    EppResource existing = ofy.load().key(Key.create(resource)).now();
+    Object existing = ofy.load().key(Key.create(resource)).now();
     if (existing != null) {
       // This will roll back the transaction and prevent duplicate history entries from being saved.
       throw new ResourceExistsException();
```
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: Good point

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 16 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: good point, missed this

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java (hunk 17 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: So I totally agree with your point here. The autorenew billing event/poll message were included here because the system required it in order to have an autorenew grace period. I think the cheapest approach here to actually handling grace period cancellations is to generate a report post-import of the TLD that flags them, and then issue refunds manually.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java (hunk 18 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: thanks!

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java (hunk 19 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: Good question. I noticed that the schema shows exDate as a nullable field, and I have encountered real deposit files with this attribute missing. With a real lack of documentation describing what a missing exDate means, it seems reasonable that the default should be one year past the domain expiration. However this has the same problem as you mentioned above, where there is no cap set at 10 years.

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java (hunk 20 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: interesting. I agree that it would be simpler to make this implicitly 1 year and eliminate the need for the field.

File: java/google/registry/rde/imports/RdeDomainImportAction.java (hunk 24 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeDomainImportAction.java]: fixed. see https://github.com/google/nomulus/pull/65

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 25 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: fixed

File: java/google/registry/rde/imports/RdeImportUtils.java [hunk 26]
```diff
@@ -80,51 +96,30 @@ public RdeImportUtils(
...
-        type, resource.getRepoId(), resource.getForeignKey());
+    return ImmutableSet.<Object>of(ForeignKeyIndex.create(resource, resource.getDeletionTime()),
+        EppResourceIndex.create(Key.create(resource)));
   }
 
   /**
-   * Imports a host from an escrow file.
+   * Imports a resource from an escrow file.
    *
    * <p>The host will only be imported if it has not been previously imported.
```
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: fixed

File: java/google/registry/rde/imports/RdeImportUtils.java (hunk 16 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/RdeImportUtils.java]: fixed

File: java/google/registry/rde/imports/XjcToDomainResourceConverter.java (hunk 19 above)
Comment (wolfgangmeyers) [inline java/google/registry/rde/imports/XjcToDomainResourceConverter.java]: Fixed

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java (hunk 22 above)
Comment (wolfgangmeyers) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: I've added tests in both

File: javatests/google/registry/rde/imports/RdeDomainImportActionTest.java (hunk 21 above)
Comment (wolfgangmeyers) [inline javatests/google/registry/rde/imports/RdeDomainImportActionTest.java]: Fixed

**Comment by wolfgangmeyers** [comment]
@nfelt I've made most of the changes you suggested, please see https://github.com/google/nomulus/pull/65 if/when you have time to review
